tapfeatures/aux_channel_negotiator.go

```diff
@@ -0,0 +1,191 @@
+	// If we fail to support a remotely required feature they are the ones
+	// to drop the connection (by returning an error right here).
+	err = checkRequiredBits(getLocalFeatureVec(), peerVec)
+	if err != nil {
+		return err
+	}
+
+	// Store this peer's features.
+	n.peerFeatures.Store(peer, peerVec)
+
+	return nil
+}
+
+// ProcessChannelReady handles the reception of the ChannelReady message, which
+// signals that a newly established channel is now ready to use. This helps us
+// correlate a peer's features with a channel outpoint
+func (n *AuxChannelNegotiator) ProcessChannelReady(cid lnwire.ChannelID,
+	peer route.Vertex) {
+
+	features, ok := n.peerFeatures.Load(peer)
+	if ok {
+		n.chanFeatures.Store(cid, features)
+	}
+}
+
+// GetReestablishFeatures is called when sending a channel_reestablish message.
+// It returns feature bits based on the specific channel identified by its
+// funding outpoint and aux channel blob.
+//
+// We're currently not using the cid and aux blob of the channel as we're
+// always using the same feature vectors regardless of the channel in question.
+func (n *AuxChannelNegotiator) GetReestablishFeatures(
+	_ lnwire.ChannelID, _ tlv.Blob) (tlv.Blob, error) {
+
+	var buf bytes.Buffer
+
+	// We could store per-channel features, but currently there's no need to
+	// support that. We report the same feature vector over all of our
+	// custom channels.
+	features := getLocalFeatureVec()
+	err := features.Encode(&buf)
+	if err != nil {
+		return nil, err
+	}
+
+	return buf.Bytes(), nil
+}
+
+// ProcessReestablishFeatures handles received channel_reestablish feature TLVs.
+// This is a blocking call - the channel link will wait for this method to
+// complete before continuing channel operations. The implementation can modify
+// aux channel behavior based on the negotiated features.
+func (n *AuxChannelNegotiator) ProcessReestablishFeatures(cid lnwire.ChannelID,
+	features tlv.Blob, _ tlv.Blob) error {
+
+	buf := bytes.NewBuffer(features)
+	peerVec := lnwire.NewRawFeatureVector()
+	err := peerVec.Decode(buf)
+	if err != nil {
+		return err
+	}
+
+	// Before we store this channel's supported features we need to first
+	// check if our required features are supported by the remote party. If
+	// a locally required feature is not supported we have to return an
+	// error and drop the connection. Whether we support all of the remote
+	// required features is a responsibility of the remote peer. If we fail
+	// to support a remotely required feature they are the ones to drop the
+	// connection (by returning an error right here).
+	err = checkRequiredBits(getLocalFeatureVec(), peerVec)
+	if err != nil {
+		return err
+	}
+
+	n.chanFeatures.Store(cid, peerVec)
+
+	return nil
+}
+
+// GetPeerFeatures returns the negotiated feature bit vector that was
+// established with the given peer.
+func (n *AuxChannelNegotiator) GetPeerFeatures(
+	peer route.Vertex) *lnwire.FeatureVector {
```

Comment:
If we're returning `nil` rather than an option, then we need to take care to always check the return value in the follow up PR. 

Otherwise, if `lnd` has restarted, but the peers haven't connected again, then we'll panic when trying to check if feature bits are present.